### PR TITLE
Add global registry to composable API contract

### PR DESCRIPTION
**Pull request by mrshenli**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #90621
* #90620
* __->__ #90579
* #90523

### CODE REVIEW

File: torch/distributed/_composable/contract.py [hunk 1]
```diff
@@ -149,4 +174,10 @@ def get_state(module: nn.Module) -> Optional[_State]:
 
     wrapper.state = get_state  # type: ignore[attr-defined]
 
+    def get_registry(module: nn.Module) -> OrderedDict[str, RegistryItem]:
```
Comment (mrshenli) [inline torch/distributed/_composable/contract.py:177]: registry order is API applied order

File: torch/distributed/_composable/contract.py [hunk 2]
```diff
@@ -19,6 +19,15 @@ class _State:
 
 
 STATE_KEY = _StateKey()
+REGISTRY_KEY = _StateKey()
+
+
+# TODO: we can add additional info to RegistryItem to share across APIs. E.g.,
+# we can add args and kwargs here, and then we can detect whether fully_shard
+# is combined with reentrant activation checkpointing and error out with a clear
+# message.
```
Comment (awgu) [inline torch/distributed/_composable/contract.py:28]: Good point!

File: torch/distributed/_composable/contract.py [hunk 3]
```diff
@@ -149,4 +174,10 @@ def get_state(module: nn.Module) -> Optional[_State]:
 
     wrapper.state = get_state  # type: ignore[attr-defined]
 
+    def get_registry(module: nn.Module) -> OrderedDict[str, RegistryItem]:
+        default_registry: Dict[str, RegistryItem] = OrderedDict()
+        return module.__dict__.setdefault(REGISTRY_KEY, default_registry)
+
+    wrapper.registry = get_registry
```
Comment (yhcharles) [inline torch/distributed/_composable/contract.py:181]: In this way, when we try to get registry, we do `replicate.registry(m)` or `fully_shard.registry(m)`, but they return the same dict. Does it make sense to move this function to a top level API, say `contract.registry(m)`?

Comment (awgu) [reply torch/distributed/_composable/contract.py:181]: This is a good point. I do not feel strongly that we must make it `contract.registry(m)`, but I agree with @yhcharles that that API would be cleaner.

**Review by awgu — COMMENTED** [review]
The usage in `test_registry` looks good to me. I think this works for what we need :)

**Review by awgu — APPROVED** [review]

**Review by yhcharles — APPROVED** [review]
Thanks! LGTM

**Comment by mrshenli** [comment]
test failure is irrelevant

```
2022-12-10T17:17:41.2691765Z ERROR ENCOUNTERED WHEN UPLOADING TO SCRIBE: Read timeout on endpoint URL: "https://lambda.us-east-1.amazonaws.com/2015-03-31/functions/gh-ci-scribe-proxy/invocations"

```

**Pull request closed without merging**